### PR TITLE
cluster manager: avoid immediate activation for dynamic inserted cluster when initialize

### PR DESCRIPTION
**Pull request by Shikugawa**
Commit Message: To resolve https://github.com/envoyproxy/envoy/issues/12670. This commit changes the cluster state change. In previous implementation, dynamically inserted cluster when initialization should be active immediately. This patch changes, from active immediately to warming -> active when initialization. 
Additional Description: 
Risk Level: Mid (core behavior change) 
Testing: Unit
Docs Changes: Needed
Release Notes: Needed
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

File: source/common/upstream/cluster_manager_impl.cc [hunk 1]
```diff
@@ -417,6 +417,11 @@ void ClusterManagerImpl::onClusterInit(Cluster& cluster) {
   // of operations here is important. We start by initializing the thread aware load balancer if
   // needed. This must happen first so cluster updates are heard first by the load balancer.
   auto cluster_data = active_clusters_.find(cluster.info()->name());
+  if (cluster_data == active_clusters_.end()) {
+    clusterWarmingToActive(cluster.info()->name());
```
[the remark this answers is not in the document]
Comment (Shikugawa) [reply source/common/upstream/cluster_manager_impl.cc:421]: We'd like to defer `clusterWarmingToActive` and `updateClusterCounts` when cluster initialize. So I put them here.

[the remark this answers is not in the document]
Comment (Shikugawa) [reply source/common/upstream/cluster_manager_impl.cc:421]: Fix this redundancy.

File: source/common/upstream/cluster_manager_impl.cc [hunk 2]
```diff
@@ -613,40 +619,39 @@ bool ClusterManagerImpl::addOrUpdateCluster(const envoy::config::cluster::v3::Cl
   //       the future we may decide to undergo a refactor to unify the logic but the effort/risk to
   //       do that right now does not seem worth it given that the logic is generally pretty clean
   //       and easy to understand.
-  const bool use_active_map =
-      init_helper_.state() != ClusterManagerInitHelper::State::AllClustersInitialized;
-  loadCluster(cluster, version_info, true, use_active_map ? active_clusters_ : warming_clusters_);
-
-  if (use_active_map) {
+  const bool all_clusters_initialized =
```
Comment (JimmyCYJ) [inline source/common/upstream/cluster_manager_impl.cc:622]: Can we add unit test to cover this change?

Comment (Shikugawa) [reply source/common/upstream/cluster_manager_impl.cc:622]: This is only to inverse cond expression to be more clear the meaning of role. So I think that this case of change is already covered with unit test.

**Review by lizan — DISMISSED** [review]
This LGTM, can you merge master? @mattklein123 for non-Tetrate pass.

**Comment by mattklein123** [comment]
Please merge main and I will take a look.

/wait

File: source/common/upstream/cluster_manager_impl.cc [hunk 3]
```diff
@@ -417,7 +417,13 @@ void ClusterManagerImpl::onClusterInit(Cluster& cluster) {
   // been setup for cross-thread updates to avoid needless updates during initialization. The order
   // of operations here is important. We start by initializing the thread aware load balancer if
   // needed. This must happen first so cluster updates are heard first by the load balancer.
-  auto cluster_data = active_clusters_.find(cluster.info()->name());
+  auto cluster_data = warming_clusters_.find(cluster.info()->name());
+  if (cluster_data != warming_clusters_.end()) {
```
Comment (mattklein123) [inline source/common/upstream/cluster_manager_impl.cc:421]: Isn't the point of this change that when this function is called the cluster is always in warming? Can you add more comments above to clarify when the cluster would not be in warming?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
At a high level LGTM with 1 question.

Also, per @lizan it would be nice to add an integration test for this behavior. Thank you!

/wait

File: source/common/upstream/cluster_manager_impl.cc (hunk 3 above)
Comment (Shikugawa) [reply source/common/upstream/cluster_manager_impl.cc:422]: The role of this change is to assure clusters which called this function should be active. Comment addition done.

Comment (mattklein123) [reply source/common/upstream/cluster_manager_impl.cc:422]: My question is how can it be that a cluster is not warming when this function is called?

/wait-any

Comment (Shikugawa) [reply source/common/upstream/cluster_manager_impl.cc:422]: [this](https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/cluster_manager_impl.cc#L294) section will be immediately active.

Comment (mattklein123) [reply source/common/upstream/cluster_manager_impl.cc:422]: OK please add more comments.

/wait

Comment (Shikugawa) [reply source/common/upstream/cluster_manager_impl.cc:424]: comment addition done

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks LGTM other than test comment. Thank you!

/wait

File: test/integration/ads_integration_test.cc [hunk 4]
```diff
@@ -1132,6 +1132,18 @@ class AdsClusterV3Test : public AdsIntegrationTest {
...
 
+TEST_P(AdsClusterV3Test, BasicClusterInitialWarming) {
+  initialize();
+  const auto cds_type_url = Config::getTypeUrl<envoy::config::cluster::v3::Cluster>(
+      envoy::config::core::v3::ApiVersion::V3);
+
+  EXPECT_TRUE(compareDiscoveryRequest(cds_type_url, "", {}, {}, {}, true));
+  sendDiscoveryResponse<envoy::config::cluster::v3::Cluster>(
+      cds_type_url, {buildCluster("cluster_0")}, {buildCluster("cluster_0")}, {}, "1", false);
+  test_server_->waitForGaugeEq("cluster_manager.warming_clusters", 0);
```
[the remark this answers is not in the document]
Comment (Shikugawa) [reply test/integration/ads_integration_test.cc:1150]: Done. like this?

[the remark this answers is not in the document]
Comment (mattklein123) [reply test/integration/ads_integration_test.cc:1150]: Not quite. Verify that the warming gauge is 1 before you send the EDS response.

/wait

[the remark this answers is not in the document]
Comment (Shikugawa) [reply test/integration/ads_integration_test.cc:1151]: done.

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Comment by mattklein123** [comment]
Merge main once https://github.com/envoyproxy/envoy/pull/13598 merges. Thanks!

/wait

**Comment by lizan** [comment]
@mattklein123 FYI a `/retest` after #13598 would suffice.

**Comment by rgs1** [comment]
FYI seeing some envoys stuck in `PRE_INITIALIZING` after syncing with upstream... so it might be this change or #13516... will know in a bit.

**Comment by rgs1** [comment]
> FYI seeing some envoys stuck in `PRE_INITIALIZING` after syncing with upstream... so it might be this change or #13516... will know in a bit.

this is when doing a hot restart

**Comment by mattklein123** [comment]
Thanks @rgs1, it wouldn't surprise me if there was some bugs here. This code is complicated. If we need to revert we can.

**Comment by rgs1** [comment]
I'll do a bisect in a bit and confirm the offending commit.

**Comment by rgs1** [comment]
Ok this is change is clean, the issue seems to be with #13516 mixed with hot restarts.

cc: @mpuncel

**Pull request merged**